client/lib/create-cached-selector/index.js

```diff
@@ -0,0 +1,102 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, forEach, some, isFunction } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/*
+ * A map that is Weak with objects but Strong with primitives
+ */
+export class MixedMap {
+	weakMap = new WeakMap();
+	map = new Map();
+
+	constructor( init ) {
+		forEach( init, ( [ key, val ] ) => {
+			this.mapForKey( key ).set( key, val );
+		} );
+	}
+
+	mapForKey = key => ( isObject( key ) ? this.weakMap : this.map );
+
+	clear() {
+		this.weakMap = new WeakMap();
+		this.map.clear();
+		return this;
+	}
+
+	set( k, v ) {
+		this.mapForKey( k ).set( k, v );
+		return this;
+	}
+
+	delete( k ) {
+		this.mapForKey( k ).delete( k );
+		return this;
+	}
+
+	get( k ) {
+		return this.mapForKey( k ).get( k );
+	}
+
+	has( k ) {
+		return this.mapForKey( k ).has( k );
+	}
+}
+
+/**
+ * Returns a selector that caches values.
+ *
+ * @param  {Function} selector      A standard selector for calculating cached result
+ * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
+ *                                    Must return an object which gets used as the first arg of the selector
+ * @return {Function}               Cached selector
+ */
+export default function createCachedSelector( { selector, getDependents } ) {
+	if ( ! isFunction( selector ) || ! isFunction( getDependents ) ) {
+		throw new TypeError(
+			'createCachedSelector: invalid arguments passed, selector and getDependents must both be functions'
+		);
+	}
+
+	const cache = new MixedMap();
+
+	const cachedSelector = function( state, ...args ) {
+		const dependents = getDependents( state, ...args );
+		const sortedDependentsArray = Object.keys( dependents )
+			.sort()
```

Comment:
this is not what you think it is, it is also not stable

Reply:
discussed offline. the keys will always be unique strings therefore [Array#sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) does what we want -- to produce the same array for the same object of dependents every time

Reply:
The enumeration order of object properties is [guaranteed](http://2ality.com/2015/10/property-traversal-order-es6.html) to be the same as the order in which the properties were created. So you don't need to sort the keys nearly as often as you might think. You still need to make cases like this behave correctly:
```js
getDependents = state => cond ? { dep1, dep2 } : { dep2, dep1 };
```

The fact that you require a particular property order, and that the keys are not used by the `createCachedSelector` for anything, suggests that the data structure you're really looking for is an array, not an object:
```js
createCachedSelector(
  ( [ dep1, dep2 ], ...args ) => combine( dep1, dep2, args ),
  ( state, ...args ) => [ dep1, dep2 ]
);
```

Or, even better in my opinion, follow the widely used `reselect` function signature:
```js
createCachedSelector(
  ( state, ...args ) => computeDep1,
  ( state, ...args ) => computeDep2,
  ( dep1, dep2, ...args ) => combine( dep1, dep2, args )
)
```
The disadvantage here is that it's hard to add a `getKey` function as another argument, if you ever needed it. And the code in #21507 suggests that you might need it.

Reply:
> The fact that you require a particular property order, and that the keys are not used by the createCachedSelector for anything, suggests that the data structure you're really looking for is an array, not an object:

They keys _are_ used!  Not in a way that is necessary, but in a way that I think makes for a nice api. The return value of `getDependents` becomes the first parameter to `selector`

So for example:
```js
getDependents = (state, siteId) => ({ site: getSite(siteId), posts: getPosts(siteId), comments: getComments(siteId)});
selector = ({site, posts, comments}) => doThings(site, posts, comments)
```
---

All that said, I may be alone in how much I love using objects for named parameters, so I'll switch to an array that may make more sense here for all the reasons you cited:
```js
getDependents = (state, siteId) => [getSite(siteId), getPosts(siteId), getComments(siteId)];
selector = ([site, posts, comments]) => doThings(site, posts, comments)
```

While the reselect api is really pleasant, I agree that we may need to sometimes use custom cache key generation which discounts that as an option.  If we were rebuilding calypso from scratch we wouldn't, but in order to easily transition code that already depends on objects as selector args then we'll need it.